### ISSUE TITLE
Should browsers do bookkeeping on DOM parts after initial parse?

### ISSUE THREAD

**Issue by iteriani** [WICG/webcomponents#1030]
See preliminary benchmark in https://playcode.io/1595899 using chrome-dev with experimental platform features on.

My numbers appear to be

cloning
8.100000023841858
getParts
1.5000000447034836
accessNodes
1.0999999791383743
operations
1.800000011920929
replaceChildren
30.5

As you can see, there is huge overhead in replaceChildren, presumably to do bookkeeping on keeping the parts up to date. Given how we expect users to use DOM parts in frameworks, we should think about whether bookkeeping is necesary in v1.

**Comment by rniwa** [comment]
What kind of bookkeepings are there?

**Comment by iteriani** [comment]
Consider the following test case.

```
const el = template`<div>{{#}}{{/#}}</div>`;
const child = template`<span>{{#}}Something here{{/#}}</span>`;
const part = el.content.getPartRoot().getParts()[0];
expect(part.getParts().length).toBe(0);
part.replaceChildren(child);
expect(part.getParts().length).toBe(1);
```

I personally would be ok with the last assertion to be 0.

**Comment by rniwa** [comment]
Why is supporting getParts() to be updated so expensive? It seems like an implementation problem to me.

**Comment by iteriani** [comment]
We’re going to look into seeing if we’re doing anything egregious. I guess
if we’re using ChildNodePart.replaceChildren with a document fragment, the
operation to update parts should be pretty cheap.

I was thinking with a normal dom element, you’d have to track its owner
part and move the related parts accordingly, but we’re not exactly doing
that.

**Comment by mfreed7** [comment]
> Why is supporting getParts() to be updated so expensive? It seems like an implementation problem to me.

I tend to agree - I’d refrain from making API decisions too early based on the early prototype. I haven’t had much time to optimize and it’s likely there’s something easy that could be fixed in replaceChildren.

That said, it’d be good to have the debate about whether tracking is *necessary*, because it certainly has non-zero overhead. Removing it will make the basic operations faster. But I don’t know if removing it will make the *usage of the API* faster, since it might require more work to be done on the framework side.

**Comment by justinfagnani** [comment]
If bookkeeping does end up being irreducibly expensive, I was definitely partial to the `template.cloneWithParts()` API before, which essentially required any bookkeeping to be done on the JS side. If you don't have uncontrolled DOM mutations, where you then have to ask for the current parts, this seems fine.

**Comment by iteriani** [comment]
Mason fixed up my benchmark a little bit, so now we're a little bit better with DOM parts being about 1.3x slower (can definitely be investigated further).

See https://playcode.io/1595899

<img width="618" alt="Screenshot 2023-09-16 at 7 03 08 AM" src="https://github.com/WICG/webcomponents/assets/3457307/01be568c-df9a-4690-b287-ad6d174e6977">

**Comment by mfreed7** [comment]
> can definitely be investigated further

…is *actively* being investigated further…
